msrestazure/tools.py

```diff
@@ -0,0 +1,87 @@
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+# sell copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+#
+# --------------------------------------------------------------------------
+
+import json
+import re
+import logging
+import time
+import uuid
+
+_LOGGER = logging.getLogger(__name__)
+
+def register_rp_hook(r, *args, **kwargs):
+    """This is a requests hook to register RP automatically.
+
+    See requests documentation for details of the signature of this function.
+    http://docs.python-requests.org/en/master/user/advanced/#event-hooks
+    """
+    if r.status_code == 409 and 'msrest' in kwargs:
+        rp_name = _check_rp_not_registered_err(r)
+        if rp_name:
+            session = kwargs['msrest']['session']
+            url_prefix = _extract_subscription_url(r.request.url)
+            _register_rp(session, url_prefix, rp_name)
+            time.sleep(10)
+            req = r.request
+            # Change the 'x-ms-client-request-id' otherwise the Azure endpoint
+            # just returns the same 409 payload without looking at the actual query
+            if 'x-ms-client-request-id' in req.headers:
+                req['x-ms-client-request-id'] = str(uuid.uuid1())
+            return session.send(req)
+
+def _check_rp_not_registered_err(response):
+    try:
+        response = json.loads(response.content.decode())
+        if response['error']['code'] == 'MissingSubscriptionRegistration':
+            match = re.match(r".*'(.*)'", response['error']['message'])
+            return match.group(1)
+    except Exception:  # pylint: disable=broad-except
+        pass
+    return None
+
+def _extract_subscription_url(url):
+    """Extract the first part of the URL, just after subscription:
+    https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/
+    """
+    match = re.match(r".*/subscriptions/[a-f0-9-]+/", url, re.IGNORECASE)
+    if not match:
+        raise ValueError("Unable to extract subscription ID from URL")
+    return match.group(0)
+
+def _register_rp(session, url_prefix, rp_name):
+    """Synchronously register the RP is paremeter."""
+    post_url = "{}providers/{}/register?api-version=2016-02-01".format(url_prefix, rp_name)
```

Comment:
I assume you have double checked that this api-version is available in sovereign clouds and stacks. I think so, but just in case. Or you can feel free to `import azure.mgmt.resource`, 99% chance, the package will be there, then you just create a client and use the sdk method.  Up to you, since you know SDK better than anyone